build/build.go

```diff
@@ -591,25 +592,79 @@ func (s *Session) buildImportPathWithSrcDir(path string, srcDir string) (*Packag
 	return pkg, archive, nil
 }
 
+const (
+	hashDebug = false
+)
+
+var (
+	gopherJsBinaryHash string
+)
+
+func hashGopherJsBinary() (string, error) {
+	if gopherJsBinaryHash != "" {
+		return gopherJsBinaryHash, nil
+	}
+
+	binHash := sha256.New()
+	binPath, err := os.Executable()
+	if err != nil {
+		return "", fmt.Errorf("could not locate GopherJS binary: %v", err)
+	}
+	binFile, err := os.Open(binPath)
+	if err != nil {
+		return "", fmt.Errorf("could not open %v: %v", binPath, err)
+	}
+	defer binFile.Close()
+	if _, err := io.Copy(binHash, binFile); err != nil {
+		return "", fmt.Errorf("failed to hash %v: %v", binPath, err)
+	}
+	gopherJsBinaryHash = fmt.Sprintf("%#x", binHash.Sum(nil))
+	return gopherJsBinaryHash, nil
+}
+
 func (s *Session) BuildPackage(pkg *PackageData) (*compiler.Archive, error) {
 	if archive, ok := s.Archives[pkg.ImportPath]; ok {
 		return archive, nil
 	}
 
+	// For non-main and test packages we build up a hash that will help
+	// determine staleness. Set hashDebug to see this in action. The format is:
+	//
+	// ## sync
+	// gopherjs binary hash: 0x519d22c6ab65a950f5b6278e4d65cb75dbd3a7eb1cf16e976a40b9f1febc0446
+	// build tags:
+	// import: internal/race
+	//   hash: 0xb966d7680c1c8ca75026f993c153aff0102dc9551f314e5352043187b5f9c9a6
+	// ...
+	//
+	// file: /home/myitcv/gos/src/sync/cond.go
+	// <file contents>
+	// N bytes
+	// ...
+
+	pkgHash := sha256.New()
+	var hw io.Writer = pkgHash
+	var hashDebugOut *bytes.Buffer
+	if hashDebug {
+		hashDebugOut = new(bytes.Buffer)
+		hw = io.MultiWriter(hashDebugOut, pkgHash)
+	}
+
 	if pkg.PkgObj != "" {
-		var fileInfo os.FileInfo
-		gopherjsBinary, err := os.Executable()
-		if err == nil {
-			fileInfo, err = os.Stat(gopherjsBinary)
-			if err == nil {
-				pkg.SrcModTime = fileInfo.ModTime()
-			}
-		}
+		fmt.Fprintf(hw, "## %v\n", pkg.ImportPath)
+
+		binHash, err := hashGopherJsBinary()
 		if err != nil {
-			os.Stderr.WriteString("Could not get GopherJS binary's modification timestamp. Please report issue.\n")
-			pkg.SrcModTime = time.Now()
+			return nil, err
 		}
 
+		fmt.Fprintf(hw, "gopherjs binary hash: %v\n", binHash)
+
+		orderedBuildTags := append([]string{}, s.options.BuildTags...)
+		sort.Strings(orderedBuildTags)
+
+		fmt.Fprintf(hw, "build tags: %v\n", strings.Join(orderedBuildTags, ","))
+
 		for _, importedPkgPath := range pkg.Imports {
```

Comment:
Is this assured that `pkg.Imports` is ordered in alphabetical order?

Reply:
Yes, this is [also assured](https://github.com/golang/go/blob/665b9b3476ad0a6dc4e578e42e6c63012e23aaa0/src/go/build/build.go#L1126-L1133).